### PR TITLE
Remove App::before from routes.php in modules to allow routes be accessible

### LINKED ISSUE #3096: Route list
You could implement a command that would list all defined routes, from plugins (already have this one but it might be integrated with a new one) and also from the whole theme or project.

### PR DESCRIPTION
**Pull request by Samuell1**
This allows/fixes routes to be visible in route:list and allows to fix route caching for plugins, backend and cms.

Before:
obrázok

After:
obrázok

Fixes #3096 , #2019

Related PR in library repository: https://github.com/octobercms/library/pull/439

### CODE REVIEW

**Comment by Samuell1** [comment]
Before this merge of this PR there needs to be added provider in because i am not sure if somebody forgot to add it when upgrading laravel, but array of providers is missing `Illuminate\Foundation\Support\Providers\RouteServiceProvider`.

**Comment by LukeTowers** [comment]
@Samuell1 what does that service provider do? It looks like it's just setting up the routes file for a Laravel application, which isn't used by us

**Comment by Samuell1** [comment]
@LukeTowers RouteServiceProvider loads and process cached routes when they exists.

**Comment by bennothommo** [comment]
@Samuell1 Will this have any ramifications with middleware? I assume the `App::before` call that's currently there, albeit being a holdover from Laravel 4, was also set up to run before any middleware?

**Comment by Samuell1** [comment]
@bennothommo I dont think so. I tried backedn and it was looking without issues.

**Comment by bennothommo** [comment]
@Samuell1 I know we've discussed this several times before on Slack, but what were the advantages of caching the routes, especially with regards to October in which everything passes through our main backend Controller class anyway in order to link correctly to installed plugins?

**Comment by Samuell1** [comment]
@bennothommo Advantage of this is to allow route:list to work and allow to cache plugin routes.

**Comment by bennothommo** [comment]
@Samuell1 I mean, performance wise? :stuck_out_tongue:

**Comment by Samuell1** [comment]
Yes, it should help with performance if you using multiple routes in plugin (for example like me). @bennothommo

**Comment by LukeTowers** [comment]
@Samuell1 Does this also support the plugin's routes.php files as well?

**Comment by Samuell1** [comment]
@LukeTowers Yes, thats why i was doing this in first place.

For example my plugin routes now are shown in route:list :
obrázok

**Comment by daftspunk** [comment]
This appears ok on the surface but I think it would have to be thoroughly tested because it involves modifying the order of events.

**Comment by Samuell1** [comment]
@daftspunk Yep, i think we can add tests for this if there is a way to test somehow routing events and what you think about this https://github.com/octobercms/october/pull/4714#issuecomment-545570609 ? Will be better to make OCMS own RouterProvider or reuse from illuminate?

**Comment by Samuell1** [comment]
Looks like Debugbar plugin that extending router trough vendor package has issue when App::before is removed from routes files. I am not sure why this happens but maybe Router is not initialized correctly?

**Comment by bennothommo** [comment]
@Samuell1 would you be willing to see if this will work with the L6 upgrade, and target your PR for that?

**Comment by LukeTowers** [comment]
@bennothommo @Samuell1 this is on my todo list to review, if you wait about a month I'll probably get to it then. There's some other changes that have to be made in connection to this.

**Comment by bennothommo** [comment]
@LukeTowers should we still target it for L6?

**Comment by Samuell1** [comment]
@LukeTowers @bennothommo i was talking long time ago with @daftspunk about this, he said he did it different or something now iam not sure if i should close this or?

**Comment by LukeTowers** [comment]
@Samuell1 leave it for now so we remember to do this. @bennothommo yeah, we'd be targetting L6 at least, if not after the L6 build gets released.

**Comment by Samuell1** [comment]
@LukeTowers @bennothommo Looks like Laravel 8 will support closures for caching too maybe it can reimplemented forward in ocms too.

**Comment by bennothommo** [comment]
@Samuell1 isn't that something that's already supported in L6? https://laravel.com/docs/6.x/cache#retrieving-items-from-the-cache

Or are we talking about something else?

**Comment by Samuell1** [comment]
@bennothommo its different Cache just saves value of that closure but for routes you need to have cached the closure to return value.

**Comment by bennothommo** [comment]
@Samuell1 ah I see. Do you have a link to some docs or an announcement about the feature?

**Comment by bennothommo** [comment]
@Samuell1 never mind, saw your message on Slack.

For reference: https://www.georgebuckingham.com/laravel-8-new-features/

**Comment by LukeTowers** [comment]
@Samuell1 is this still something that you want merged?

**Comment by bennothommo** [comment]
This will need to be adjusted to implement changes done with the L6 upgrade - mainly changing `{slug}` to `{slug?}`.

File: modules/backend/routes.php [hunk 1]
```diff
@@ -3,30 +3,29 @@
...
-     */
-    Route::any(Config::get('cms.backendUri', 'backend'), 'Backend\Classes\BackendController@run')->middleware('web');
+/*
+* Other pages
+*/
+Route::group([
+        'middleware' => ['web'],
+        'prefix' => Config::get('cms.backendUri', 'backend')
+    ], function () {
+        Route::any('{slug}', 'Backend\Classes\BackendController@run')->where('slug', '(.*)?');
```
Comment (bennothommo) [inline modules/backend/routes.php:19]:
```suggestion
        Route::any('{slug?}', 'Backend\Classes\BackendController@run')->where('slug', '(.*)?');
```

File: modules/cms/routes.php [hunk 2]
```diff
@@ -3,20 +3,19 @@
...
-    /*
-     * Extensibility
-     */
-    Event::fire('cms.route');
-});
+/*
+* The CMS module intercepts all URLs that were not
+* handled by the back-end modules.
+*/
+Route::any('{slug}', 'Cms\Classes\CmsController@run')->where('slug', '(.*)?')->middleware('web');
```
Comment (bennothommo) [inline modules/cms/routes.php:16]:
```suggestion
Route::any('{slug?}', 'Cms\Classes\CmsController@run')->where('slug', '(.*)?')->middleware('web');
```

**Review by bennothommo — CHANGES_REQUESTED** [review]

**Comment by Samuell1** [comment]
@bennothommo yes but @daftspunk had some ideas how to make it work without effecting order of route initialization but iam not sure now if he did something more.

**Comment by daftspunk** [comment]
Hi @Samuell1. I can confirm it will be fixed in the new version of October CMS coming out soon. It will include route caching. Closing for now

**Pull request closed without merging**